nepalingo-web/src/components/StreakContext.tsx

```diff
@@ -0,0 +1,78 @@
+import React, {
+  createContext,
+  useContext,
+  useState,
+  useEffect,
+  ReactNode,
+} from "react";
+import supabase from "./userAuth/supabaseClient";
+import { useAuth } from "./userAuth/AuthContext";
+
+interface StreakContextProps {
+  currentStreak: number;
+  longestStreak: number;
+  updateStreak: () => void;
+}
+
+export const StreakContext = createContext<StreakContextProps>({
+  currentStreak: 0,
+  longestStreak: 0,
+  updateStreak: () => {},
+});
+
+export const StreakProvider = ({ children }: { children: ReactNode }) => {
+  const { user } = useAuth();
+  const [currentStreak, setCurrentStreak] = useState(1);
+  const [longestStreak, setLongestStreak] = useState(1);
+
+  useEffect(() => {
+    const fetchStreakData = async () => {
+      if (user) {
+        const { data } = await supabase
+          .from("user_daily_streaks")
+          .select("*")
+          .eq("user_id", user.id)
+          .single();
+
+        if (data) {
+          setCurrentStreak(data.current_streak);
+          setLongestStreak(data.longest_streak);
+        }
+      }
+    };
+
+    fetchStreakData();
+  }, [user]);
+
+  const updateStreak = async () => {
+    const currentDate = new Date().toISOString().split("T")[0]; // Get current date
+    const lastUpdateDate = localStorage.getItem("lastUpdateDate");
+
+    if (user) {
+      const { data } = await supabase.rpc("update_streak", {
+        user_id: user.id,
+      });
```

Comment:
Is the `update_streak` rpc created somewhere in supabase? I wasn't able to find it

Reply:
it is in supabase databases. sanskar created a database table to store the streaks

Reply:
![CleanShot 2024-07-17 at 23 21 11@2x](https://github.com/user-attachments/assets/a31a1baf-f4f1-4feb-b431-6efb5dac6742)
Yeah I'm not seeing the streaks actually updating and am getting a 404 for that function. I'm so confused 😂

Reply:
u don't see the streak cuz it's 0 days lol. Come back again tomorrow to see 1 day streak

Reply:
and the error.....hm... I thought I fixed it. Btw the streak is updated through the updateStreak function in the StreakContext. I did test it, it worked tho. still... I have some doubts

Reply:
am I not able to upsert tables from the context cuz we need functions directly in the database? well....that explains it!